KGL_Gene
========

C++ Population Genome Comparison Toolkit
----------------------------------------

This software is under an MIT license, you may use or modify any part of the code for any purpose.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OR GUARANTEE OF ANY KIND.

A toolkit that uses Fasta/Gff3/VCF files to analyze Insert/Deletes (Indels) and Single Nucleotide Polymorphisms (SNP)
variants in genome populations.

The toolkit was developed specifically to analyze mutations in blood cell 
membrane proteins (stevor, rifin, var and cleft) generated by the malaria parasite Plasmodium
Falciparum. It has been extensively tested on this organism. The software also supports other 
organisms including Homo Sapiens (you) and has been tested on the latest (Apr 2020) GRCh38.p13 genome definition. 
   

This software currently resides in the GitHub repository:
https://github.com/kellerberrin/KGL_Gene

Software requirements:

Target operating system is Linux (the software is being developed using VS Code/Clion under Mint 19).

Compiler must support C++ 17 standards, g++ -v >= 7.0 (the software is developed using g++ version 7.5)

cmake -version >= 3.8

Hardware requirements:

A multi-core (minimum 4 cores) 64 bit Intel/AMD PC or laptop with at least 16GB memory (>= 32GB best)
and 100GB free disk space.

The fast install procedure.
---------------------------

There is no fast install procedure (but watch this space).

The software is still under (very) active development and no stable release yet exists.

The software uses Boost and the C++ sequence analysis library Seqan.

If you have any questions then please email me (James McCulloch) at: james.duncan.mcculloch@gmail.com.

Or alternatively, register an issue on GitHub.


