KGL_Gene
========

C++ Population Genome Comparison Toolkit
----------------------------------------

This software is under an MIT license, you may use or modify any part of the code for any purpose.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OR GUARANTEE OF ANY KIND.

A toolkit that uses Fasta/Gff3/VCF files to analyze Insert/Deletes (Indels) and Single Nucleotide Polymorphisms (SNP)
variants in genome populations.

The toolkit was developed specifically to analyze mutations in blood cell 
membrane proteins (stevor, rifin, var and cleft) generated by the malaria parasite Plasmodium
Falciparum. It has been extensively tested on this organism. The software also supports other 
organisms including Homo Sapiens (you) and has been used to analyze on the latest (Feb 2021) 
GRCh38.p13 Homo Sapien genome variant databases. For best performance, 1TB ecc ram is required 
and 100 execution threads. 
This hardware requirement is now common on corporate servers or large workstations
  

The software currently resides in the GitHub repository:
https://github.com/kellerberrin/KGL_Gene

Software requirements:

Target operating system is Linux and has been developed on a workstation 
running the Mint 20.04 Linux distribution. The software should compile and be usable under 
other recent Linux distributions but this is not tested.
The software is written to the C++ 17 standard using the Clion IDE.
A switch to C++ 20 is planned soon. I wish to thank JetBrains for providing me with 
a free Clion license to develop this open source software.

The software is still under (very) active development and no stable release yet exists.

Hardware requirements:

The software is compute and memory intensive and requires a 1/10/100 class workstation; 
1TB ecc memory, 10TB fast (nvme) disk storage and 100 execution threads. 


If you have any questions then please email me
(James McCulloch) at: james.duncan.mcculloch@gmail.com.


